### DISCUSSION TITLE
Which tools should we provide modules for in core AA?

### DISCUSSION THREAD

**Discussion by jooh** [automaticanalysis/automaticanalysis#227] [Ideas]
I would like to propose that we radically cut the number of external toolboxes that we offer modules for in core AA.

It's always nice to add new modules to AA, but new content comes with a cost in terms of user support, documentation, and software testing. The aa_modules folder is already huge and difficult to keep track of even for core developers.

To my mind, the two core weaknesses of AA at the moment are 1) absent or outdated documentation, 2) no software tests. I think it will be difficult if not impossible to document and test all of the current aa_modules. And we add new ones with almost every pull request.

The alternative to including modules and dependencies in core AA is to distribute modules in separate projects and repos, which can work as plugins to extend core AA. This is the approach I took with my modules for MVPA based on the [pilab](https://github.com/jooh/pilab) toolbox. These modules are available in their own [pilab_aa](https://github.com/jooh/pilab_aa) repository. This is good for AA core development, because it's now my problem to document, test, and support these modules. It also speeds up core AA development (you don't have to wait until my pilab modules work until a pull request can be merged).

This issue has been on my mind for a while, but I was reminded of it now because I noticed the new PR from @tiborauer includes modules that use the CosmoMVPA toolbox. This is the *third* MVPA toolbox that AA provides some functionality for, to my knowledge. We already bundle @alexvicegrab 's MVPAA under aa_toolbox (although I'm not sure if anyone is using it at the moment), and there's at least one module for the DMLT toolbox.

To my mind, the core dependency for AA is SPM. You would struggle to run AA without it, so I think any module that only requires SPM is fair to include in core AA. Beyond this, I think it's worthwhile to provide modules that additionally require the other main neuro packages - FSL and Freesurfer.

But at the risk of sounding radical, I wonder if we could just take all modules that have additional dependencies beyond this and branch them off to their own 'plugin' projects. That would mean cutting modules that use

* GLMDenoise
* ANTS
* LoAD
* DMLT
* MVPAA
* ANLM
* DKI / DTInl
* tsdiffana

Perhaps worth having a discussion about this in the next haangout. The 'kitchen sink' development model we are currently following with AA has its strengths, but I wonder if the model really is sustainable long term. Certainly I don't think there's much hope of substantially improving the documentation or tests as it stands.

Let me know your thoughts.

**Comment by jpeelle** [comment]
I like this idea, and it also makes clear who is supporting the extra
functionality. My only objection to the list is I think tsdiffana could
stay as it’s not really under active development and would be relatively
easy to fix/update. I would, though, add the Lateralisation Index (LI)
toolbox which we have recently been playing around with. I agree this is a
good topic for the next haangout.

**Comment by alexvicegrab** [comment]
Happy for you to retire MVPAA, as I've not worked on it for a long time,
and there are probably better maintained tools now.

On 26 July 2018 at 22:24, Jonathan Peelle <notifications@github.com> wrote:

> I like this idea, and it also makes clear who is supporting the extra
> functionality. My only objection to the list is I think tsdiffana could
> stay as it’s not really under active development and would be relatively
> easy to fix/update. I would, though, add the Lateralisation Index (LI)
> toolbox which we have recently been playing around with. I agree this is a
> good topic for the next haangout.
>
> On Thu, Jul 26, 2018 at 12:50 AM Johan Carlin <notifications@github.com>
> wrote:
>
> > I would like to propose that we radically cut the number of external
> > toolboxes that we offer modules for in core AA.
> >
> > It's always nice to add new modules to AA, but new content comes with a
> > cost in terms of user support, documentation, and software testing. The
> > aa_modules folder is already huge and difficult to keep track of even for
> > core developers.
> >
> > To my mind, the two core weaknesses of AA at the moment are 1) absent or
> > outdated documentation, 2) no software tests. I think it will be
> difficult
> > if not impossible to document and test all of the current aa_modules. And
> > we add new ones with almost every pull request.
> >
> > The alternative to including modules and dependencies in core AA is to
> > distribute modules in separate projects and repos, which can work as
> > plugins to extend core AA. This is the approach I took with my modules
> for
> > MVPA based on the pilab <https://github.com/jooh/pilab> toolbox. These
> > modules are available in their own pilab_aa
> > <https://github.com/jooh/pilab_aa> repository. This is good for AA core
> > development, because it's now my problem to document, test, and support
> > these modules. It also speeds up core AA development (you don't have to
> > wait until my pilab modules work until a pull request can be merged).
> >
> > This issue has been on my mind for a while, but I was reminded of it now
> > because I noticed the new PR from @tiborauer
> > <https://github.com/tiborauer> includes modules that use the CosmoMVPA
> > toolbox. This is the *third* MVPA toolbox that AA provides some
> > functionality for, to my knowledge. We already bundle @alexvicegrab
> > <https://github.com/alexvicegrab> 's MVPAA under aa_toolbox (although
> I'm
> > not sure if anyone is using it at the moment), and there's at least one
> > module for the DMLT toolbox.
> >
> > To my mind, the core dependency for AA is SPM. You would struggle to run
> > AA without it, so I think any module that only requires SPM is fair to
> > include in core AA. Beyond this, I think it's worthwhile to provide
> modules
> > that additionally require the other main neuro packages - FSL and
> > Freesurfer.
> >
> > But at the risk of sounding radical, I wonder if we could just take all
> > modules that have additional dependencies beyond this and branch them off
> > to their own 'plugin' projects. That would mean cutting modules that use
> >
> > - GLMDenoise
> > - ANTS
> > - LoAD
> > - DMLT
> > - MVPAA
> > - ANLM
> > - DKI / DTInl
> > - tsdiffana
> >
> > Perhaps worth having a discussion about this in the next haangout. The
> > 'kitchen sink' development model we are currently following with AA has
> its
> > strengths, but I wonder if the model really is sustainable long term.
> > Certainly I don't think there's much hope of substantially improving the
> > documentation or tests as it stands.
> >
> > Let me know your thoughts.
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/rhodricusack/automaticanalysis/issues/153>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AAGtPW6HGvarz0F_
> HiaHayIZocJbBTmXks5uKZ7agaJpZM4VhmWe>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rhodricusack/automaticanalysis/issues/153#issuecomment-408239850>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABgK2mOqM50598WzXtRP1vhqmTjIXRESks5uKjOhgaJpZM4VhmWe>
> .
>



-- 
Vicente Grabovetsky, Alejandro (Sasha)
Data Scientist, PhD
uk.linkedin.com/in/sashavg

**Comment by tiborauer** [comment]
I also support the distributed coding, however, I am not sure how it works within the GitHub ecosystem. E.g. can we have repos within repos on GitHub and/or in the local installation? We also have to work out the 'plugin' mechanism, so that aa can load them during aa_init.

The list above:
- I would, perhaps, also keep DKI and tsdiffana, because they are strongly modified/upgraded versions in comparison with their original code base. But they are independent modules, so I am not so strong on this.
- DTInl is not in the repo any more.
- I would also retire MVPAA, if not maintained/used.

**Comment by jpeelle** [comment]
I took Johan’s suggestion to be separate repos that users would have to
make sure were in their path. We could provide an example of this and I
think that would be sufficient. In the README for each add-on repo we can
explain this. It makes methods sections slightly mroe complicated but in
the long run seems more sustainable.

**Comment by tiborauer** [comment]
I prefer not to rely on user too much. :) I think a solution similar to that of SPM could work well:

- An add-on folder should be set up. If we cannot put git repos within aa, then the location of the folder outside aa should be specified in aap_parameters_defaults. We can, then, remove all the entries corresponding to external tools (LI, BrainWavelet, etc.) from it.
- Add-ons should have some meta-data (in XML or JSON) containing the location of the toolbox they are using, so they can be added to the path (similarly to the current solution)
- aa_init could go through the add-ons and toolboxes based on the meta-data and add required folders to the path (while making sure that .git folders are ignored).

**Comment by tiborauer** [comment]
There is one more issue with the distribuited model: At the moment the 'official' aa repo is in Rhodri's account. However, if add-on repos will sit in individual accounts, then they may be hard to find by users. I suggest creating an "automaticanalysis" 'organisation account' and put the 'official' core aa repo and the future add-on repos there.

**Comment by tiborauer** [comment]
We could call add-ons aadd-ons. :D

**Comment by jooh** [comment]
My thought was indeed to organise the plugins as completely independent repositories. We can include links to 'recommended' plugins in the core AA README to make sure people know about what's out there. But in principle it wouldn't be for us to regulate what plugins people choose to offer for AA... An AA organisation might make sense, but I guess that's @rhodricusack's call. It's anyway somewhat a separate issue.

In terms of implementation, I think Tibor's suggestion is exactly what I had in mind. I think we will find that the module list and the parameters xml files will both get dramatically shorter. The only additional comment there is that the plugin XML file also needs to have a field for adding shell variables.

I'm not at all precious about what exact content we break out into plugins and what stays in core AA. It sounds like we are converging on

*Keep*
* tsdiffana

*As plugins*
* GLMDenoise
* ANTS
* LoAD
* DMLT
* ANLM
* DKI (NB any customisations we've made to DKI could go into that plugin's repo instead)
* LI
* BrainWavelet

*Retire*
* MVPAA

Does that sound ok with everyone?

**Comment by tiborauer** [comment]
Switching to add-ons and where to store them may or may not be a separate issue. I would rather set up an effcient ecosystem now than patching it later.

As a part of it, we also have to work out the add-on 'interface' (i.e. the XML). I like the idea of specifying requirements, such as MATLAB path to the toolbox, linux path to external tools (e.g. for FSL and freesurfer), environmental variables, etc.; so that aa_init can set them accordingly.

_Regarding modules:_
- DTI and DKI belongs together. SPM does not support either and we use the (customised) DKI toolbox for DTI, as well. Another thing to consider is that DTI/DKI is also integrated into the configuration level (domains for diffusion_session and for AP-PA subsessions). Finally, DTI/DKI processing involves a lot of FSL code, which is also not part of the SPM, thus may not be considered as 'core aa' by definition.
- What shall we do with FSL and Freesurfer-related modules? Freesurfer is quite standalone at the moment and cold be separated easily. FSL may require a bit more work and consideration, but it may be still feasible.
- Do we care about dependency between add-ons (e.g. DTI/DKI depends on FSL? Do we want to model it (e.g. corresponding field in the add-on interface)?

**Comment by tiborauer** [comment]
_I would also add these to the add-on list:_
- aamod_LouvainCluster.m, aamod_LOO_LouvainCluster.xml
- marsbar
- MPM (using hMRI toolbox)
- PEwarp
- MFP (switch in aamod_realign(unwarp) has to be modified)
- GIFT
- tissue_wavelets (MATLAB's Wavelet Toolbox)
